### PR TITLE
Refactor GPU Reader Op

### PR DESCRIPTION
**Pull request by Kh4L**
Signed-off-by: Serge Panev <spanev@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/reader/reader_op.h [hunk 1]
```diff
@@ -197,60 +197,28 @@ class DataReader : public Operator<Backend> {
...
-      Operator<Backend>::Run(ws);
-      loader_->ReturnTensor(prefetched_batch_[samples_processed_]);
+    if (prefetch_error_) {
+      DALI_FAIL("Prefetching failed: " + dali::string(*prefetch_error_));
     }
 
-    // lock, check if batch is finished, notify
-    {
-      std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
+    producer_.notify_one();
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/reader_op.h]: I'd like to know a little bit of the motivation of the refactoring, and whether this code is unit tested

File: dali/pipeline/operators/reader/reader_op.h [hunk 2]
```diff
@@ -197,60 +197,28 @@ class DataReader : public Operator<Backend> {
...
-        // signal the other workers we're ready
-        prefetch_ready_workers_ = true;
+    // grab the actual prefetching lock
+    std::unique_lock<std::mutex> prefetch_lock(prefetch_access_mutex_);
 
-        producer_.notify_one();
-      }
-    }
+    // Wait until prefetch is ready
+    consumer_.wait(prefetch_lock, [&]() { return prefetched_batch_ready_; });
```
[the remark this answers is not in the document]
Comment (Kh4L) [reply dali/pipeline/operators/reader/reader_op.h]: Done

File: dali/pipeline/operators/reader/reader_op.h (hunk 1 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/reader_op.h]: The thread safety needed by the CPU version of the reader is not required by the GPU one
Tested in a python pipeline but not unit tested in C++

File: dali/pipeline/operators/reader/reader_op.h [hunk 3]
```diff
@@ -197,60 +197,28 @@ class DataReader : public Operator<Backend> {
...
-      // if we've consumed all samples in this batch, reset state and stop
-      if (samples_processed_.load() == Operator<Backend>::batch_size_) {
-        prefetch_ready_workers_ = false;
-        prefetched_batch_ready_ = false;
-        producer_.notify_one();
-        samples_processed_ = 0;
-        batch_stop_ = true;
-      }
-      return;
+    for (size_t i = 0; i < prefetched_batch_.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (Kh4L) [reply dali/pipeline/operators/reader/reader_op.h]: Done

**Pull request merged**